packages/plugin/vite/src/util/package.ts

```diff
@@ -0,0 +1,120 @@
+import fs from 'node:fs';
+import path from 'node:path';
+
+import type { PackageJsonManifest } from './packageJson';
+
+export interface Dependency {
+  name: string;
+  path: SourceAndDestination;
+  dependencies: Dependency[];
+}
+
+export interface SourceAndDestination {
+  src: string;
+  dest: string;
+}
+
+function isRootPath(dir: string) {
+  // *unix or Windows root path
+  return dir === '/' || /^[a-zA-Z]:\\$/i.test(dir);
+}
+
+export async function isDirectory(p: string): Promise<boolean> {
+  try {
+    const stat = await fs.promises.stat(p);
+    return stat.isDirectory();
+  } catch {
+    return false;
+  }
+}
+
+export async function lookupNodeModulesPaths(root: string, paths: string[] = []): Promise<string[]> {
+  if (!root) return paths;
+  if (!path.isAbsolute(root)) return paths;
+
+  const p = path.join(root, 'node_modules');
+
+  if (await isDirectory(p)) {
+    paths = paths.concat(p);
+  }
+  root = path.join(root, '..');
+
+  return isRootPath(root) ? paths : await lookupNodeModulesPaths(root, paths);
+}
+
+export async function readPackageJson(root = process.cwd()): Promise<PackageJsonManifest> {
+  const packageJsonPath = path.join(root, 'package.json');
+  try {
+    const packageJsonStr = await fs.promises.readFile(packageJsonPath, 'utf8');
+    return JSON.parse(packageJsonStr);
+  } catch (error) {
+    console.error(`parse ${packageJsonPath} failed: ${error}`);
+    throw error;
+  }
+}
+
+export async function resolveDependencies(root: string) {
+  const rootDependencies = Object.keys((await readPackageJson(root)).dependencies || {});
+  const resolve = async (prePath: string, dependencies: string[], collected: Map<string, Dependency> = new Map()) =>
+    await Promise.all(
+      dependencies.map(async (name) => {
+        let curPath = prePath,
+          depPath = null,
+          packageJson = null;
+        while (!packageJson && !isRootPath(curPath)) {
+          const allNodeModules = await lookupNodeModulesPaths(curPath);
+
+          for (const nodeModules of allNodeModules) {
+            depPath = path.join(nodeModules, name);
+            if (fs.existsSync(depPath)) break;
```

Comment:
If we end up using `fs-extra`, we can replace this sync call by:

```suggestion
            if (await fs.pathExists(depPath)) break;
```